### DISCUSSION TITLE
addding an NLOHMANN_DEFINE_TYPE_READONLY macro

### DISCUSSION THREAD

**Discussion by e2sn4uudk** [nlohmann/json#3118] [Ideas]
This additional macro allows easy serialization of objects with const members, and sometimes serialization is all a project needs (https://github.com/karlkurzer/json/pull/1/commits/7ecc6fdb52072bc30fdfeb3391cd0bf27fe4b317)

What do you think? Or does that already exist and I just haven't seen it?

**Comment by nlohmann** [comment]
No `NLOHMANN_DEFINE_TYPE_READONLY`-style macro exists in the current macro set — the existing macros all generate a `from_json` that assigns directly to member variables via `get_to`, which requires non-const, assignable members. Supporting const members would need a constructor-based `from_json` (returning a newly constructed `Type` rather than mutating one in place), which is a reasonable and fairly self-contained addition for types that only need serialization plus immutable construction.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*

**Comment by gregmarr** [comment]
@nlohmann I don't know for sure because the fork is gone, but I believe this is looking for `NLOHMANN_DEFINE_TYPE_[NON_]INTRUSIVE_ONLY_SERIALIZE` which does only `to_json()`, not `from_json()`.
